crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -1115,6 +1308,82 @@ pub(crate) async fn deserialize_move_struct(
     Ok((struct_tag, move_struct))
 }
 
+/// Check that the cursors, if provided, have the same checkpoint_viewed_at.
+pub(crate) fn validate_cursor_consistency(
+    after: Option<&Cursor>,
+    before: Option<&Cursor>,
+) -> Result<Option<u64>, Error> {
+    match (after, before) {
+        (Some(after_cursor), Some(before_cursor)) => {
+            if after_cursor.checkpoint_viewed_at == before_cursor.checkpoint_viewed_at {
+                Ok(Some(after_cursor.checkpoint_viewed_at))
+            } else {
+                Err(Error::Client(
+                    "Cursors are inconsistent and cannot be used together in the same query."
+                        .to_string(),
+                ))
+            }
+        }
+        (Some(cursor), None) | (None, Some(cursor)) => Ok(Some(cursor.checkpoint_viewed_at)),
+        (None, None) => Ok(None),
+    }
+}
+
+fn objects_query(filter: &ObjectFilter, lhs: i64, rhs: i64) -> RawQuery {
+    // Construct the filtered inner query - apply the same filtering criteria to both
+    // objects_snapshot and objects_history tables.
+    let mut snapshot_objs = query!(r#"SELECT * FROM objects_snapshot"#);
+    snapshot_objs = filter.apply(snapshot_objs);
```

Comment:
nit: Can you mention here that it's important that the `ObjectFilter` is applied first, because of how its `OR` filters are structured?